yjit/src/codegen.rs

```diff
@@ -1429,70 +1431,70 @@ gen_newrange(jitstate_t *jit, ctx_t *ctx, codeblock_t *cb)
 
     return YJIT_KEEP_COMPILING;
 }
+*/
 
-static void
-guard_object_is_heap(codeblock_t *cb, x86opnd_t object_opnd, ctx_t *ctx, uint8_t *side_exit)
+fn guard_object_is_heap(cb: &mut CodeBlock, object_opnd: X86Opnd, ctx: &mut Context, side_exit: CodePtr)
 {
     add_comment(cb, "guard object is heap");
 
     // Test that the object is not an immediate
-    test(cb, object_opnd, imm_opnd(RUBY_IMMEDIATE_MASK));
+    test(cb, object_opnd, imm_opnd(RUBY_IMMEDIATE_MASK as i64));
     jnz_ptr(cb, side_exit);
 
     // Test that the object is not false or nil
-    cmp(cb, object_opnd, imm_opnd(Qnil));
-    RUBY_ASSERT(Qfalse < Qnil);
+    let VALUE(qnilval) = Qnil;
+    cmp(cb, object_opnd, imm_opnd(qnilval as i64));
     jbe_ptr(cb, side_exit);
 }
 
-static inline void
-guard_object_is_array(codeblock_t *cb, x86opnd_t object_opnd, x86opnd_t flags_opnd, ctx_t *ctx, uint8_t *side_exit)
+fn guard_object_is_array(cb: &mut CodeBlock, object_opnd: X86Opnd, flags_opnd: X86Opnd, ctx: &mut Context, side_exit: CodePtr)
 {
     add_comment(cb, "guard object is array");
 
     // Pull out the type mask
-    mov(cb, flags_opnd, member_opnd(object_opnd, struct RBasic, flags));
-    and(cb, flags_opnd, imm_opnd(RUBY_T_MASK));
+    mov(cb, flags_opnd, mem_opnd(8 * SIZEOF_VALUE as u8, object_opnd, RUBY_OFFSET_RBASIC_FLAGS));
+    and(cb, flags_opnd, imm_opnd(RUBY_T_MASK as i64));
 
     // Compare the result with T_ARRAY
-    cmp(cb, flags_opnd, imm_opnd(T_ARRAY));
+    cmp(cb, flags_opnd, imm_opnd(RUBY_T_ARRAY as i64));
     jne_ptr(cb, side_exit);
 }
 
 // push enough nils onto the stack to fill out an array
-static codegen_status_t
-gen_expandarray(jitstate_t *jit, ctx_t *ctx, codeblock_t *cb)
+fn gen_expandarray(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock, ocb: &mut OutlinedCb) -> CodegenStatus
 {
-    int flag = (int) jit_get_arg(jit, 1);
+    let flag = jit_get_arg(jit, 1);
+    let VALUE(flag_value) = flag;
 
     // If this instruction has the splat flag, then bail out.
-    if (flag & 0x01) {
-        gen_counter_incr!(cb, expandarray_splat);
-        return YJIT_CANT_COMPILE;
+    if flag_value & 0x01 == 0x1 {
+        incr_counter!(expandarray_splat);
+        return CantCompile;
     }
 
     // If this instruction has the postarg flag, then bail out.
-    if (flag & 0x02) {
-        gen_counter_incr!(cb, expandarray_postarg);
-        return YJIT_CANT_COMPILE;
+    if flag_value & 0x02 == 0x2 {
+        incr_counter!(expandarray_postarg);
+        return CantCompile;
     }
 
-    uint8_t *side_exit = get_side_exit(jit, ocb, ctx);
+    let side_exit = get_side_exit(jit, ocb, ctx);
 
     // num is the number of requested values. If there aren't enough in the
     // array then we're going to push on nils.
-    int num = (int)jit_get_arg(jit, 0);
-    val_type_t array_type = ctx_get_opnd_type(ctx, OPND_STACK(0));
-    x86opnd_t array_opnd = ctx_stack_pop(ctx, 1);
+    let num = jit_get_arg(jit, 0);
+    let VALUE(num_value) = num;
```

Comment:
We should define helper methods for VALUE imo, like `VALUE.as_isize()` or `VALUE.as_usize()`, which the appropriate checks inside using into().